docs/gen-ai/gen-ai-events.md

```diff
@@ -0,0 +1,379 @@
+For example, when using OpenAI client libraries to communicate with Mistral, the `gen_ai.system`
+is set to `openai` based on the instrumentation's best knowledge.
+
+For custom model, a custom friendly name SHOULD be used.
+If none of these options apply, the `gen_ai.system` SHOULD be set to `_OTHER`.
+
+
+
+`gen_ai.system` has the following list of well-known values. If one of them applies, then the respective value MUST be used; otherwise, a custom value MAY be used.
+
+| Value  | Description | Stability |
+|---|---|---|
+| `anthropic` | Anthropic | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| `cohere` | Cohere | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| `openai` | OpenAI | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| `vertex_ai` | Vertex AI | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+
+
+
+<!-- markdownlint-restore -->
+<!-- prettier-ignore-end -->
+<!-- END AUTOGENERATED TEXT -->
+<!-- endsemconv -->
+
+## System message
+
+This event describes the instructions passed to the GenAI model.
+It's NOT RECOMMENDED to report this event when capturing content is disabled, unless there are additional system-specific fields
+instrumentation adds to the event body.
+
+The event name MUST be `gen_ai.system.message`.
+
+| Body Field | Type | Description | Examples | Requirement Level |
+|---|---|---|---|---|
+| `content` | string | The contents of the system message. | `"You're a friendly bot that answers questions about OpenTelemetry."` | `Opt-In` |
+
+## User message
+
+This event describes the prompt message specified by the user.
+
+The event name MUST be `gen_ai.user.message`.
+
+| Body Field | Type | Description | Examples | Requirement Level |
+|---|---|---|---|---|
+| `content` | `AnyValue` | The contents of the user message. | `What telemetry is reported by OpenAI instrumentations?` | `Opt-In` |
+
+## Assistant message
+
+This event describes the assistant message.
+
+The event name MUST be `gen_ai.assistant.message`.
+
+| Body Field   | Type                           | Description                            | Examples                                        | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) |
+|--------------|--------------------------------|----------------------------------------|-------------------------------------------------|-------------------|
+| `content`    | `AnyValue`                     | The contents of the assistant message. | `Spans, events, metrics defined by the GenAI semantic conventions.` | `Opt-In` |
+| `tool_calls` | [ToolCall](#toolcall-object)[] | The tool calls generated by the model, such as function calls. | `[{"id":"call_mszuSIzqtI65i1wAUOE8w5H4", "function":{"name":"get_link_to_otel_semconv", "arguments":"{\"semconv\":\"gen_ai\"}"}, "type":"function"}]` | `Conditionally Required`: if available |
+
+### `ToolCall` object
```

Comment:
copying some folks from last LLM call @nirga @karthikscale3 @AmanAgarwal041 @ThomasVitale

unless I am mistaken, this is a big feature change of the spec. Before, there was to my knowledge no attributes or other developments on tool calls. I'm not sure how compatible this is across the board, but I noticed for example openllmetry has an open issue about [anthropic tool calls](https://github.com/traceloop/openllmetry/issues/879).

Aside, I think the description could be fleshed out @lmolkova to specify what features this PR supports (modeling tool calls is a very big deal for interop, and somewhat independent from being represented in log events). This is a suggestion you can feel free to ignore if you think it is implied.

In summary, while I understand not all backends will be able to support log events soon, and may shrug at this PR thinking they will just convert back to span events. Even if that's the case, you should still look closely at the model here. It isn't just changing to log events, it is also fleshing out features not currently tracked in a specific issue (notably tool calls).

Reply:
tool calls are hidden  in underspecified prompt and completion events.

https://platform.openai.com/docs/guides/function-calling

they are part of gen-ai responses and input sent by users. This PR describes the structure and clarifies the content of the events. You can't address it without specifying all the basic pieces. 

I.e. this PR does not add any new big feature.

Reply:
The reason I raised this comment is besides [one comment](https://github.com/open-telemetry/semantic-conventions/pull/980#discussion_r1594321576) from @langchain4j in an early draft of this PR, I don't see any discussion about tool calling. I took time to make this thread because it isn't obvious in the title or description that tool call modeling is in the same scope as the mechanics of span event -> log events. Almost all discussion was on the latter point.

While it may indeed be a small change, so far I don't see any one discussing it, much less looking up cohere, anthropic or other non-openai change. There are a number of fields around tool calls which may be brittle if not looked at independent of the other parts of this change.

Two suggestions FWIW:
* fork off tool call specification into a separate issue PR and remove it from this one
* redo the description and on the next call get a diverse opinion on how the changes here apply

The latter is tricky because of the same reason the main PR is tricky. As the api isn't available, people need to theorize about the change vs practice it.

My 2p, and last opinion until someone else pipes up. Thanks so far!

Reply:
the title of this PR says "introduce per-message structured GenAI events". 

The tool calls is just on of them, I can't separate it from this PR because then GenAI conventions will not be complete.

I.e. previously we assumed tool calls to be in the prompt and completion span event. Now they what? should not be captured?

I will keep tool calls inside the scope of this PR. This is an experimental spec and I'm sure we will need to specify and update the structure further.

Reply:
> While it may indeed be a small change, so far I don't see any one discussing it, much less looking up cohere, anthropic or other non-openai change. There are a number of fields around tool calls which may be brittle if not looked at independent of the other parts of this change.

if you have some specific proposal on what needs to be added to better support non-openai and what currently in this PR does not work for them, please bring specific suggestions up.

Reply:
To reinforce this -

This PR should be evaluated based on the proposed EVENT being defined.  This is replacing attributes with an Event, and that event should serve the same purpose.   This *is* a feature change.  If you disagree with the shape of the Event being proposed - comment as such.

I think we (OpenTelemetry) fragment problems a bit too often to the point where we're no longer solving an issue, but solving a fragment of an issue and causing larger ones.  If you have a proposal for a better shaped event, please make it.  If you think ToolCalls shouldn't exist, please say so, and provide justification.

TL;DR; please make this clear as a reviewer of this PR.  What is the current issue with ToolCalls?  That should be an obvious statement on this line of code for maintainers to understand the concern and address it.  I'm unable to read your statement and figure out what you want to see in this design.

Reply:
First, the reason I pinged folks on this topic was an attempt to get current implementors to consider this issue, as they'll have experience. There are many already modeling span attributes or otherwise to represent tool events. These may or may not be easy to index. I'm not sure which are instrumenting the tool calls, and if that's done via spans, if they are adding their version of tool call ID to the LLM spans, or if they are making child spans for each tool call etc. I was hoping that people writing instrumentation can mention what if any concerns they have. I think those working on this every day will see what I'm saying and decide if this is an issue or not.

I'm being asked to present a problem on my own, so no issues doing so. Here goes. I mentioned Anthropic because I noticed openllmetry having an open issue for it. It is a good example to elaborate why it should be a separate topic vs being a sidecar of something to switch from span events to log events.

In [anthropic docs](https://docs.anthropic.com/en/docs/build-with-claude/tool-use#handling-tool-use-and-tool-result-content-blocks), they say this:

> Unlike APIs that separate tool use or use special roles like tool or function, Anthropic’s API integrates tools directly into the user and assistant message structure.

1. In our spec, we hint there is a `tool` role (e.g. `gen_ai.tool.message`), but that's not the case in anthropic. Is that ok? What does cohere do?

While there is an example, fields or events themselves have no description of what a tool call is or any implications of it. In anthropic "All tools are user-provided", so there are no server-side tools. This means it is a very good question, if these should be spans or not.. more on this later.

2. Should there be a description of tool calls in general, or if they are user provided or not? While we do hint that they are always user provided in the examples, I believe some description should exist outside the examples.

At least in anthropic, tool calls are logic defined in the same process as what calls the LLM.  At that point you are told to "Execute the actual tool code client-side." So what we call a `gen_ai.response.finish_reasons`? In the example, we hint at `tool_calls` (`tool_use` in anthropic) but it isn't listed as a choice

3. consider adding `tool_calls` to finish_reasons or explain why not

Next, so at that point you invoke a call. Now, if instrumenting, perhaps some instrument these tool calls also as a span. Right now, in the examples we say "Application executes the tool call. Application may create span which is not covered by this semantic convention." I don't personally think this should be only defined in an example. Also, knowing how any of the existing implementations work is helpful to guide this part of the spec. Do they set the span's parent as the LLM span? Do they add an attribute of `tool_call_id` to explain latency or errors?

4. Should we make any relationship note about instrumentation of user-provided tool calls? For example, the span should be a child of the LLM one and it should have an attribute of the `tool_call_id`. I think saying that spans should be parented to the LLM if done, is reasonable to say (in the spec, not just hinted in an example)

Next, right now there is no mention about tool errors, a type of message or anything. In Anthropic there is an `is_error` Do current implementors not care about this part, because for example everything is done sequentially and the error would bubble up into the calling span, or capturing it would be too expensive? If we don't care about errors, saying why raises less questions.

5. Should there be a way to know if a tool call was an error? How would we represent that?

I stopped at five. While I'm happy to hear @lmolkova's take on this, I really do want to hear what folks who are currently implementing this feel about it also. If I fail at getting feedback from implementors so be it, it was worth a shot!

Reply:
@jsuereth to your question also:

> This PR should be evaluated based on the proposed EVENT being defined. This is replacing attributes with an Event, and that event should serve the same purpose. This is a feature change. If you disagree with the shape of the Event being proposed - comment as such.

"This is replacing attributes with an Event, and that event should serve the same purpose" and basically that's my whole point. This PR is not limiting itself to replacement of existing model. I goes into modeling of something not yet modeled (tool calls). Before this change, we had basic prompt/completion span events. The change as done now extends that to tool calls, etc.

I feel this point is being treated as a nit, or draconian. The request to at least say that this is also getting into new models, if not branch off a new issue. I honestly wonder how my feedback is resisted so much? @jack-berg can you help me understand from a different POV, why I am incorrect to raise issues like this? I really want to understand and use my time wisely. I tend to spend a lot of hours on non-technical back and forth and really want to get to a point where I can get more done. Raise an issue, help solve it, and done. Less about me doing something wrong for raising it.

Reply:
I'll address your technical comments one-by-one. To offer some POV in general: your feedback was not specific or actionable so far. You've dismissed my explanation that tool calls were already inside existing attributes and this PR makes things obvious instead of implicit. And that tool calls therefor cannot be separated for it. 

I'll be happy to hear domain-specific feedback based on facts with specific suggestions. I appreciate it's being left as a separate comments that we can effectively discuss and resolve independently.

I.e. the problem is the form of your feedback and lack of clarity in it.

Reply:
I believe I addressed your questions in separate comments on this PR. As you may notice many of them seem to be out of scope of this PR and represent important work we need to do in llm semconv. 

I prefer us to focus on this work rather than nitpicking on PR descriptions. If you want to volunteer and help us do this work, you're more than welcome.

Reply:
I don't appreciate this comment being dismissed as nitpicking, as I made it exactly on the line where the content was discussed. https://github.com/open-telemetry/semantic-conventions/pull/980#discussion_r1714651646

I tagged people who are implementing this to please take a close look, only the part addressed to you personally was maybe find a way to make more visible some changes that didn't have much or any discussion yet. Then, I was told to do that for them, and I did. I still think we should be more interested in the implementations vs telling me that calling attention to things is a nit or non-work.

Reply:
@nirga @karthikscale3 @AmanAgarwal041 @ThomasVitale I did my best to raise issues on the content of things around tool calling. I will not be making comments on the follow-up threads here, as I have to respect my boundaries. I have unsubscribed from this issue.

I would encourage you to actually try this PR in javascript if you can't yet in python due to missing event api. I know everyone is very busy, so probably not all of you will be able to. If you decide to approve without trying it, that's your prerogative also. That said, you have the most experience, many of you dealing with multiple non-openai implementations. I think things are best if you can participate. Good luck.

Reply:
to summarize my understanding (whether you see it or not): you wanted other people to take a look and share the feedback. You don't want things to be sent as PRs without wide discussion. You are not interested in discussing technical details and expect others to discuss them. 
I assumed you wanted me to take some actions (remove `ToolCall` type description?), but it seems the main intention was to raise awareness,. Good to know.

The best discussion indeed happens when people implement things. Since there were no concerns/discussions on this in this PR (or in LLM calls), I assume there are no major concerns. 

This spec stays experimental for the time being and, as in other specs, we'll adjust things based on the implementation feedback. Implementation in a diverse set of languages and specific products is a prereq to stability.
If anyone discovers anything that feels wrong, confusing, suboptimal, etc, please share now or later once you're implementing it.